README.md
biblio-sici
===========

Perl module ("Biblio::SICI") for creating, manipulating and parsing SICIs.

A "Serial Item and Contribution Identifier" (SICI) is a code (ANSI/NISO standard Z39.56) used to uniquely identify specific volumes, articles or other identifiable parts of a periodical.


Biblio-SICI/Changes

{{$NEXT}}

0.04      2014-01-14 22:02:39 Europe/Berlin
	- Fix compatibility with Moo <= 1.000008 which did not allow simple
		scalars as default values

0.03      2014-01-12 17:00:16 Europe/Berlin
	- Get rid of File::Find::Rule dependency;
		use in-core File::Find

0.01      2014-01-11 21:09:48 Europe/Berlin
	- Initial release



Biblio-SICI/lib/Biblio/SICI.pm

package Biblio::SICI;

# ABSTRACT: Provides methods for assembling, parsing, manipulating and serialising SICIs

use strict;
use warnings;
use 5.010001;

use Moo;
use Sub::Quote;

use Biblio::SICI::ItemSegment;
use Biblio::SICI::ContributionSegment;
use Biblio::SICI::ControlSegment;

use Biblio::SICI::Util qw( calculate_check_char );

=pod

=encoding utf-8

=head1 SYNOPSIS

  use Biblio::SICI;

  my $sici = Biblio::SICI->new()->parse($someSICI);

  # or

  my $sici2 = Biblio::SICI->new();

  $sici2->item->issn('0361-526X');

  # ... setting more data attributes ...
  
  if ( $sici2->is_valid ) {
      say $sici->to_string;
  }

=head1 WARNING

This software is currently considered BETA. Things should work as intended
and documented but if you use it you should test your own software extensively
after any update to a new release of Biblio::SICI since both API and behaviour
might have changed.

=head1 DESCRIPTION

A "Serial Item and Contribution Identifier" (SICI) is a code (ANSI/NISO
standard Z39.56) used to uniquely identify specific volumes, articles 
or other identifiable parts of a periodical.

This module provides methods for assembling, parsing, manipulating and
serialising SICIs.

Both internal implementation and public API are currently considered BETA
and may change without warning in a future release. For more information 
on this have a look at the L<TODO section|/TODO> below.

=head1 CONFIGURATION

You may specify the following option when instantiating a SICI object
(i.e., when calling the "new()" constructor):

=over 4

=item C<mode>

Can be either C<strict> or C<lax>.

C<strict> mode means that any operation that gets called with an 
invalid (according to the standard) value for an attribute will
C<die()>.

C<lax> mode means that any value is accepted and that you can use
the C<is_valid()> and C<list_problems()> methods to analyze the 
object state.

=back

=head1 ATTRIBUTES

=over 4

=item C<item>

An instance of L<Biblio::SICI::ItemSegment>; this segment contains
information about the serial item itself.

=cut

has 'item' => (
	is   => 'ro',
	lazy => 1,
	isa => quote_sub(q{ die unless ( defined $_[0] and $_[0]->isa('Biblio::SICI::ItemSegment') ) }),
	builder =>
		quote_sub(q{ my ($self) = @_; return Biblio::SICI::ItemSegment->new( _sici => $self ); }),
	init_arg => undef,
);

=item C<contribution>

An instance of L<Biblio::SICI::ContributionSegment>; this segment
contains information about an individual contribution to the whole
item, e.g. an article in a journal issue.

=cut

has 'contribution' => (
	is   => 'ro',
	lazy => 1,
	isa  => quote_sub(
		q{ die unless ( defined $_[0] and $_[0]->isa('Biblio::SICI::ContributionSegment') ) }),
	builder => quote_sub(
		q{ my ($self) = @_; return Biblio::SICI::ContributionSegment->new( _sici => $self ); }),
	init_arg => undef,
);

=item C<control>

An instance of L<Biblio::SICI::ControlSegment>; this segment
contains some meta-information about the thing described by the
SICI and about the SICI itself.

=cut

has 'control' => (
	is   => 'ro',
	lazy => 1,
	isa  => quote_sub(
		q{ my ($val) = @_; die unless ( defined $val and $val->isa('Biblio::SICI::ControlSegment') ) }
	),
	builder => quote_sub(
		q{ my ($self) = @_; return Biblio::SICI::ControlSegment->new( _sici => $self ); }),
	init_arg => undef,
);

=item C<mode>

Describes wether the object enforces strict conformance to
the standard or not.
Can be set to either C<strict> or C<lax>.
This attribute is the only one that can be specified directly
in the call of the constructor.

Please keep in mind that changing the value does B<not> mean
that the attributes already present are re-checked!

=cut

has 'mode' => (
	is       => 'rw',
	isa      => quote_sub(q{ my ($val) = @_; die unless ( $val eq 'strict' or $val eq 'lax' ) }),
	required => 1,
	coerce   => sub {
		my ($val) = @_;
		$val = join( '', split( " ", lc($val) ) );
		return $val if ( $val eq 'strict' or $val eq 'lax' );
		return 'lax';
	},
	default => quote_sub(q{ "lax" }),
);

=item C<parsedString>

Returns the original string that was passed to the C<parse()> 
method or C<undef> if C<parse> was not called before.

=cut

has 'parsedString' => ( is => 'rwp', init_arg => undef, );

=back

=head1 METHODS

=over 4

=item C<parse>( STRING )

Tries to disassemble a string passed to it into the various
components of a SICI.

If I<strict> mode is enabled, it will C<die()> if either the
string cannot be parsed or if no valid SICI can be derived 
from the string.

If I<lax> mode is enabled, it returns a list of three values:

The first value is C<undef> if parsing the string failed, or
C<0> if the string could be parsed but the SICI is invalid, or 
C<1> if a valid SICI was found.

The second value is also C<undef> if parsing the string failed,
or C<0> if the string could be parsed but serializing the SICI
does not result in the exact same string or C<1> if we get a full 
round-trip.

The third value is a (possibly empty) array ref with a list of 
problems the parser detected. B<Please note:> these problems
are distinct from those reported by the C<list_problems> method
and can be retrieved again later.

=cut

sub parse {
	my ( $self, $string ) = @_;
	my $strictMode = $self->mode() eq 'strict' ? 1 : 0;

	if ( defined $string ) {
		$string =~ s/\r/ /go;
		$string =~ s/\n/ /go;
		$string = join( '', split( " ", $string ) );
	}

	unless ($string) {
		$strictMode ? die 'no string to parse' : return ( undef, undef, ['no string to parse'] );
	}
	$self->_set_parsedString($string);

	my $checkChar = '';
	if ( $string =~ /;([0-9])(?:-.)?\Z/ ) {
		if ( "$1" ne "2" ) {
			$strictMode
				? die 'unsupported SICI version'
				: return ( undef, undef, ['unsupported SICI version'] );
		}
		else {
			$self->control()->version(2);
			if ( $string =~ s/;2-(.)\Z// ) {
				$checkChar = $1;
			}
		}
	}

	my $parserProblems = [];

	my @chars = split( //, $string );
	my $tmp = '';
	while ( exists( $chars[0] ) and $chars[0] !~ /[(<]/ ) {
		$tmp .= shift @chars;
	}

	if ( $tmp and exists( $chars[0] ) ) {
		if ( $chars[0] eq '(' ) {

			# warn 'ISSN candidate: ' . $tmp;
			$self->item()->issn($tmp);
			shift @chars;
		}
		elsif ( $chars[0] eq '<' ) {
			push @{$parserProblems}, "item information missing";

			# warn 'Missing item info';
			# warn 'Enumeration candidate: ' . $tmp;
			if ( $tmp =~ /\A([A-Z0-9\/]+):([A-Z0-9\/]+)(?::([+*]))?\Z/ ) {
				$self->item()->volume($1);
				$self->item()->issue($2);
				$self->item()->supplOrIdx($3) if $3;
			}
			elsif ($tmp) {
				$self->item()->enumeration($tmp);
			}
			shift @chars;
			goto CONTRIB;
		}
		else {
			$strictMode ? die 'unparsable string' : return ( undef, undef, ['unparsable string'] );
		}
	} ## end if ( $tmp and exists( ...))
	else {
		$strictMode ? die 'unparsable string' : return ( undef, undef, ['unparsable string'] );
	}

	$tmp = '';
	while ( exists( $chars[0] ) and $chars[0] ne ')' ) {
		$tmp .= shift @chars;
	}
	if ( $tmp and exists( $chars[0] ) and $chars[0] eq ')' ) {

		# warn 'Chronology candidate: ' . $tmp;
		$self->item()->chronology($tmp);
		shift @chars;
	}
	elsif ( exists( $chars[0] ) and $chars[0] eq ')' ) {
		shift @chars;
	}
	else {
		$strictMode ? die 'unparsable string' : return ( undef, undef, ['unparsable string'] );
	}

	$tmp = '';
	while ( exists( $chars[0] ) and $chars[0] ne '<' ) {
		$tmp .= shift @chars;
	}
	if ( $tmp and exists( $chars[0] ) and $chars[0] eq '<' ) {

		# warn 'Enumeration candidate: ' . $tmp;
		if ( $tmp =~ /\A([A-Z0-9\/]+):([A-Z0-9\/]+)(?::([+*]))?\Z/ ) {
			$self->item()->volume($1);
			$self->item()->issue($2);
			$self->item()->supplOrIdx($3) if $3;
		}
		elsif ($tmp) {
			$self->item()->enumeration($tmp);
		}
		shift @chars;
	}
	elsif ( exists( $chars[0] ) and $chars[0] eq '<' ) {
		shift @chars;
	}
	else {
		$strictMode ? die 'unparsable string' : return ( undef, undef, ['unparsable string'] );
	}

	CONTRIB:
	$tmp = '';
	while ( exists( $chars[0] ) and $chars[0] ne '>' ) {
		$tmp .= shift @chars;
	}
	if ( $tmp and exists( $chars[0] ) and $chars[0] eq '>' ) {

		# warn 'Contribution candidate: ' . $tmp;
		if ( $tmp =~ /\A::(.+)\Z/ ) {
			$self->contribution()->localNumber($1);
		}
		elsif ( $tmp =~ /\A:([^:]+)(?::(.+))?\Z/ ) {
			$self->contribution()->titleCode($1);
			$self->contribution()->localNumber($2) if $2;
		}
		elsif ( $tmp =~ /\A([^:]+):([^:]+)(?::(.+))?\Z/ ) {
			$self->contribution()->location($1);
			$self->contribution()->titleCode($2);
			$self->contribution()->localNumber($3) if $3;
		}
		else {
			$self->contribution()->location($tmp);
		}
		shift @chars;
	}
	elsif ( exists( $chars[0] ) and $chars[0] eq '>' ) {
		shift @chars;
	}
	else {
		$strictMode ? die 'unparsable string' : return ( undef, undef, ['unparsable string'] );
	}

	my $csi = '';
	if ( exists( $chars[0] ) ) {
		$csi = shift @chars;
	}

	if ( exists( $chars[0] ) and $chars[0] eq '.' ) {
		shift @chars;
	}
	elsif ( exists( $chars[0] ) ) {
		shift(@chars) while ( $chars[0] ne '.' );
	}

	if ( exists( $chars[0] ) ) {
		$self->control()->dpi( shift @chars );
	}

	if ( exists( $chars[0] ) and $chars[0] eq '.' ) {
		shift @chars;
	}
	elsif ( exists( $chars[0] ) ) {
		shift(@chars) while ( $chars[0] ne '.' );
	}

	if ( exists( $chars[0] ) and exists( $chars[1] ) ) {
		$self->control()->mfi( join( '', splice( @chars, 0, 2 ) ) );
	}

	my $isValid = $self->is_valid();
	if ( $strictMode && !$isValid ) {
		die 'parsing failed: invalid SICI';
	}

	if ( $checkChar ne $self->checkchar() ) {
		push @{$parserProblems}, "wrong check char; was '$checkChar', should have been '$1'";
	}
	if ( $checkChar !~ /\A[0-9A-Z#]\Z/ ) {
		push @{$parserProblems}, 'invalid original check char';
	}

	if ( $self->control()->csi() ne $csi ) {
		push @{$parserProblems}, 'wrong csi in string input';
	}

	my $roundTrip = ( $self->parsedString() eq $self->to_string() ? 1 : 0 );
	return ( $isValid, $roundTrip, $parserProblems );
} ## end sub parse

=item C<to_string>

Serializes the object to a string using the separator characters
specified in the standard and returns it together with the check
character appended.

Does B<not> verify if the resulting SICI is valid!

=cut

sub to_string {
	my $self = shift;

	my $str = $self->_to_string();
	my $cs  = calculate_check_char($str);

	return $str . $cs;
}

sub _to_string {
	my $self = shift;

	my $item    = $self->item()->to_string();
	my $contrib = $self->contribution()->to_string();
	my $control = $self->control()->to_string();

	return sprintf( '%s<%s>%s-', $item, $contrib, $control );
}

=item STRING C<checkchar>()

Stringifies the object first, then calculates (and returns) 
the checksum character.
Does B<not> check, if the stringified SICI is valid!

=cut

sub checkchar {
	my $self = shift;

	my $siciAsString = $self->_to_string();

	return calculate_check_char($siciAsString);
}

=item C<reset>()

Resets all attributes to their default values.

Does not modify the C<mode> attribute.

=cut

sub reset {
	my $self = shift;
	$self->item()->reset();
	$self->contribution()->reset();
	$self->control()->reset();
	return;
}

=item BOOL C<is_valid>()

Determines if all of the attribute values stored in the object
are valid and returns either a I<true> or I<false> value.

B<TODO> check if any required information is missing!

=cut

sub is_valid {
	my $self = shift;

	my $itemIsValid    = $self->item()->is_valid();
	my $contribIsValid = $self->contribution()->is_valid();
	my $controlIsValid = $self->control()->is_valid();

	if ( $itemIsValid && $contribIsValid && $controlIsValid ) {
		return 1;
	}

	return 0;
}

=item HASHREF C<list_problems>()

Returns either a hash of hashes of arrays containing the 
problems that were found when setting the various attributes
of the SICI segments or C<undef> if there are no problems.

The first hash level is indexed by the three SICI segments:
I<item>, I<contribution>, and/or I<control>.

The level below is indexed by the attribute names (cf. the
docs of the segment modules).

For every attribute the third level contains an array reference
with descriptive messages.

  {
      'contribution' => {
          'titleCode' => [
              'contains more than 6 characters',
          ],
      },
  };

B<TODO> check for meta problems (e.g. missing attributes).

=cut

sub list_problems {
	my $self = shift;

	my $hasProblems = 0;
	my %problems    = ();
	if ( not $self->item()->is_valid() ) {
		$hasProblems++;
		$problems{'item'} = { $self->item()->list_problems() };
	}
	if ( not $self->contribution()->is_valid() ) {
		$hasProblems++;
		$problems{'contribution'} = { $self->contribution()->list_problems() };
	}
	if ( not $self->control()->is_valid() ) {
		$hasProblems++;
		$problems{'control'} = { $self->control()->list_problems() };
	}

	if ($hasProblems) {
		return %problems;
	}
	return;
}

1;

=back

=head1 TODO

The parsing of SICI strings sort-of works but I need to find
out more about how the code copes with real world SICIs (i.e. 
especially those that are slightly malformed or invalid).

It would probably make for a better programming style if I were
using real type specifications for the attributes. On the other
hand doing so would make the module employ overly strict checks
when dealing with imperfect SICIs. 
Since type checks in Moo (or Moose) know nothing about the object, 
the only other solution I can think of would be using objects of 
type C<Biblio::SICI> act as frontend for instances of either 
I<Biblio::SICI::Strict> or I<Biblio::SICI::Lax>.
This would require two separate sets of type definitions and make 
everything more complicated - and I am not sure if would provide 
us with a better way to handle and report formal problems.

That said, I´m also not particularly happy with how 
C<list_problems()> works right now and I´d be grateful for any
suggestions for improvements (or for positive feedback if it works
for you).

Also for now only problems with the available data are detected
while missing or inconsistend data is not checked for.

And of course we need a more comprehensive test suite.

=head1 SEE ALSO

L<https://en.wikipedia.org/wiki/Serial_Item_and_Contribution_Identifier>

=cut


Biblio-SICI/lib/Biblio/SICI/ControlSegment.pm

package Biblio::SICI::ControlSegment;

# ABSTRACT: The control segment of a SICI

use strict;
use warnings;
use 5.010001;

use Moo;
use Sub::Quote;

use Biblio::SICI;
with 'Biblio::SICI::Role::ValidSegment', 'Biblio::SICI::Role::RecursiveLink';

=pod

=encoding utf-8

=head1 SYNOPSIS

  my $sici = Biblio::SICI->new();

  $sici->control->csi(2);

=head1 DESCRIPTION

I<Please note:> You are expected to not directly instantiate objects of this class!

The control segment of a SICI describes various aspects of the I<thing> referenced
by the SICI using pre-defined codes.
The segment also contains some meta-information about the SICI itself. 

=head1 ATTRIBUTES

For each attribute, clearer ("clear_") and predicate ("has_") methods
are provided.

=over 4

=item C<dpi>

The I<Derivative Part Identifier> tells us, what kind of I<thing> is described
by the SICI. It can take one of four different values:

B<0> => SICI describes Serial Item or Contribution itself
B<1> => SICI describes ToC of Serial Item or Contribution
B<2> => SICI describes Index of Serial Item or Contribution
B<3> => SICI describes Abstract of Serial Item or Contribution

The default value is B<0>.

=cut

has 'dpi' => (
	is        => 'rw', lazy => 1,
	trigger   => 1,
	default   => quote_sub(q{ 0 }),
	predicate => 1,
	clearer   => 1,
);

sub _trigger_dpi {
	my ( $self, $newVal ) = @_;

	if ( not( "$newVal" eq "0" or "$newVal" eq "1" or "$newVal" eq "2" or "$newVal" eq "3" ) ) {
		$self->log_problem_on( 'dpi' => ['value not in allowed range (0|1|2|3)'] );
	}
	else {
		$self->clear_problem_on('dpi');
	}

	return;
}

=item C<mfi>

The I<Medium / Format Identifier> can take one of these codes:

B<CD> => Computer-readable optical media (CD-ROM)
B<CF> => Computer-readable magnetic disk media
B<CO> => Online (remote)
B<CT> => Computer-readable magnetic tape media
B<HD> => Microfilm
B<HE> => Microfiche
B<SC> => Sound recording
B<TB> => Braille
B<TH> => Printed text, hardbound
B<TL> => Printed text, looseleaf
B<TS> => Printed text, softcover
B<TX> => Printed text
B<VX> => Video recording
B<ZN> => Multiple physical forms
B<ZU> => Physical form unknown
B<ZZ> => Other physical form 

The default value is B<ZU>.

=cut

has 'mfi' => (
	is        => 'rw', lazy => 1,
	trigger   => 1,
	default   => quote_sub(q{ "ZU" }),
	predicate => 1,
	clearer   => 1,
);

sub _trigger_mfi {
	my ( $self, $newVal ) = @_;

	if ( $newVal !~ /\A(?:C[DFOT]|H[DE]|SC|T[BHLSX]|VX|Z[NUZ])\Z/ ) {
		$self->log_problem_on( 'mfi' => ['unknown identifier'] );
	}
	else {
		$self->clear_problem_on('mfi');
	}

	return;
}

=item C<version>

The number of the standards version to which the SICI should adhere.
The default is B<2> (which means Z39.56-1996), since that is also 
the only currently supported version.

=cut

has 'version' => (
	is        => 'rw', lazy => 1,
	trigger   => 1,
	default   => quote_sub(q{ 2 }),
	predicate => 1,
	clearer   => 1,
);

sub _trigger_version {
	my ( $self, $newVal ) = @_;

	if ( $newVal != 2 ) {
		$self->log_problem( 'version' => ['unsupported version number (i.e. not "2")'] );
	}
	else {
		$self->clear_problem_on('version');
	}

	return;
}

=back

=head1 METHODS

=over 4

=item C<csi>

The I<Code Structure Identifier> tells something about which parts of the
SICI carry values.
It can take one of three values:

B<1> => SICI for Serial Item
B<2> => SICI for Serial Contribution
B<3> => SICI for Serial Contribution "with obscure numbering"

This method automatically derives the correct value from the presence
of the respective data elements in the item and contribution segments.
If no data is present in the contribution segment the final default is B<1>.

=cut

sub csi {
	my $self = shift;

	if ( $self->_sici()->contribution()->has_localNumber() ) {
		return 3;
	}

	if (   $self->_sici()->contribution()->has_location()
		or $self->_sici()->contribution()->has_titleCode() )
	{
		return 2;
	}

	return 1;
}

=item STRING C<to_string>()

Returns a stringified representation of the data in the
control segment.

Please note that the check digit is I<not> considered to be a
part of the control segment (but the "-" preceding it in the SICI
string is).

=cut

sub to_string {
	my $self = shift;

	# Every attribute in this class has a default value
	return sprintf( '%s.%s.%s;%s', $self->csi(), $self->dpi(), $self->mfi(), $self->version() );
}

=item C<reset>()

Resets all attributes to their default values.

=cut

sub reset {
	my $self = shift;
	$self->clear_dpi();
	$self->clear_problem_on('dpi');
	$self->clear_mfi();
	$self->clear_problem_on('mfi');
	$self->clear_version();
	$self->clear_problem_on('version');
	return;
}

=item BOOL C<is_valid>()

Checks if the data for the control segment conforms
to the standard.

=back

=head1 SEE ALSO

L<Biblio::SICI::Role::ValidSegment>

=cut

1;


Biblio-SICI/lib/Biblio/SICI/Role/ValidSegment.pm
package Biblio::SICI::Role::ValidSegment;

# ABSTRACT: Role to provide common validation error handling functionality to the segments

use strict;
use warnings;

use Moo::Role;

=pod

=encoding utf-8

=head1 DESCRIPTION

A role that provides an attribute and some methods used for validation error
handling functionality in the three SICI segments.

=head1 ATTRIBUTES

=over 4

=item validationErrors

Stores the problem reports that are gathered when any attribute is set.
Do not access this directly - use the methods below.

=cut

has validationErrors => ( is => 'lazy', );

sub _build_validationErrors {
	return {};
}

=back

=head1 METHODS

=over 4

=item C<log_problem_on>( STRING, ARRAYREF )

Stores an array ref of problem descriptions for a particular attribute.

=cut

sub log_problem_on {
	my ( $self, $attr, $desc ) = @_;

	if ( defined($attr) and $attr and defined($desc) and $desc ) {
		$self->validationErrors->{$attr} = $desc;
	}
	return;
}

=item C<clear_problem_on>( STRING )

Removes the problem report for a particular attribute.

=cut

sub clear_problem_on {
	my ( $self, $attr ) = @_;

	delete $self->validationErrors->{$attr} if exists $self->validationErrors->{$attr};
	return;
}

=item HASH C<list_problems>()

Returns a hash structure with all error reports.

=cut

sub list_problems {
	my $self = shift;
	return %{ $self->validationErrors };
}

=item BOOL C<is_valid>()

Checks if any problem reports were recorded for any attribute.
If yes, returns I<FALSE>; otherwise, returns I<TRUE>.

Does not yet do extended verification; e.g. checking if all
required data is present or if there are any conflicts within
the stored SICI data!

=cut

sub is_valid {
	my $self    = shift;
	my $invalid = 0;

	my %problems = %{ $self->validationErrors };

	foreach my $attr ( keys %problems ) {
		if ( exists $problems{$attr} and defined $problems{$attr} and $problems{$attr} ne '' ) {
			$invalid++;
		}
	}
	return 0 if $invalid;
	return 1;
}

=back

=cut

1;


Biblio-SICI/lib/Biblio/SICI/ContributionSegment.pm

package Biblio::SICI::ContributionSegment;

# ABSTRACT: The contribution segment of a SICI

use strict;
use warnings;
use 5.010001;

use Moo;
use Sub::Quote;

use Biblio::SICI;
with 'Biblio::SICI::Role::ValidSegment', 'Biblio::SICI::Role::RecursiveLink';

use Biblio::SICI::Util ();

=pod

=encoding utf-8

=head1 SYNOPSIS

  my $sici = Biblio::SICI->new();

  $sici->contribution->location('6-21');

=head1 DESCRIPTION

I<Please note:> You are expected to not directly instantiate objects of this class!

The contribution segment of a SICI contains information about a contribution 
to an item, e.g. about an article published in a journal issue.
A SICI can be valid without containing data on a contribution so this segment may
be empty / unused.

For further information please have a look at the standard.

=head1 ATTRIBUTES

For each attribute, clearer ("clear_") and predicate ("has_") methods
are provided.

=over 4

=item C<location>

The location of the contribution within the item. A typical example
would be the range of pages an article was printed on.

=cut

has 'location' => ( is => 'rw', trigger => 1, predicate => 1, clearer => 1, );

sub _trigger_location {
	my ( $self, $newVal ) = @_;

	if ( $newVal !~ /\A${Biblio::SICI::Util::TITLE_CODE}+\Z/ ) {
		$self->log_problem_on( 'location' => ['contains invalid characters'] );
	}
	else {
		$self->clear_problem_on('location');
	}

	return;
}

=item C<titleCode>

A special code of no more than 6 characters, derived from the title 
of the contribution.
Rules for the construction of the code are provided in the standard.

=cut

has 'titleCode' => ( is => 'rw', trigger => 1, predicate => 1, clearer => 1, );

sub _trigger_titleCode {
	my ( $self, $newVal ) = @_;
	my @problems = ();

	if ( length $newVal > 6 ) {
		push @problems, 'contains more than 6 characters';
	}

	if ( $newVal !~ /\A${Biblio::SICI::Util::TITLE_CODE}+\Z/ ) {
		push @problems, 'contains invalid characters';
	}

	if (@problems) {
		$self->log_problem_on( titleCode => \@problems );
	}
	else {
		$self->clear_problem_on('titleCode');
	}

	return;
}

=item C<localNumber>

An opaque number that somehow describes the contribution
in the context of its item.

=back

=cut

has 'localNumber' => ( is => 'rw', trigger => 1, predicate => 1, clearer => 1, );

sub _trigger_localNumber {
	my ( $self, $newVal ) = @_;

	if ( $newVal !~ /\A${Biblio::SICI::Util::TITLE_CODE}+\Z/ ) {
		$self->log_problem_on( 'location' => ['contains invalid characters'] );
	}
	else {
		$self->clear_problem_on('location');
	}

	return;
}

=head1 METHODS

=over 4

=item STRING C<to_string>()

Returns a stringified representation of the data in the
contribution segment.

=cut

sub to_string {
	my $self = shift;
	my $str  = '';

	if ( $self->has_location() ) {
		$str .= $self->location();
	}

	if ( $self->has_titleCode() ) {
		$str .= ':' . $self->titleCode();
	}

	if ( $self->has_localNumber() ) {
		if ( $self->has_location() or $self->has_titleCode() ) {
			$str .= ':' . $self->localNumber();
		}
		else {
			$str .= '::' . $self->localNumber();
		}
	}

	return $str;
}

=item C<reset>()

Resets all attributes to their default values.

Resetting the contribution segment also resets the value of the
C<csi> attribute in the control segment!  

=cut

sub reset {
	my $self = shift;
	$self->clear_location();
	$self->clear_problem_on('location');
	$self->clear_titleCode();
	$self->clear_problem_on('titleCode');
	$self->clear_localNumber();
	$self->clear_problem_on('localNumber');

	return;
}

=item BOOL C<is_valid>()

Checks if the data for the contribution segment conforms
to the standard.

=back

=head1 SEE ALSO

L<Biblio::SICI::Role::ValidSegment>

=cut

1;


Biblio-SICI/lib/Biblio/SICI/Util.pm

package Biblio::SICI::Util;

# ABSTRACT: Utility functions

use strict;
use warnings;
use 5.010001;

BEGIN {
	$Biblio::SICI::Util::TITLE_CODE = qr/[0-9A-Z&´*\\\{\}\(\)\[\],\@\$=!#%.+?";\/^\`~_|-]/;
}

use Exporter 'import';
our @EXPORT_OK = qw( calculate_check_char titleCode_from_title );

use Try::Tiny;

=pod

=encoding utf-8

=head1 SYNOPSIS

  use Biblio::Sici;
  use Biblio::SICI::Util qw( titleCode_from_title calculate_check_char );

  my $sici = Biblio::Sici->new();

  ...

  if ( my $code = titleCode_from_title($title) ) {
      $sici->contribution->titleCode($code);
  }

=head1 DESCRIPTION

This module provides some utility functions which are useful when
working with SICI.

None of them are exported by default.

=head1 FUNCTIONS

=over 4

=item STRING C<titleCode_from_title>( STRING )

Tries to derive the C<titleCode> (cf. the contribution segment) 
from a string (i.e.: the title of the contribution) passed to it. 

Since the rules for the construction of the C<titleCode> are
quite complex, this method probably won´t generate the correct
code if the title contains elements other than regular english 
words (like mathematical or chemical symbols or characters from
a non-ascii alphabet).

If the method is able to generate a code it will be returned.
Otherwise the return value is C<undef>.

In order to have a better chance of generating a standard-conformant 
code the modules L<Text::Unidecode> and L<Text::Undiacritic> are 
C<require>d.
If they cannot be found, warnings are emitted but a code
might nonetheless be generated. However, if the title is in
a language other than english, the probability of generating
the correct code sinks even more.

=cut

sub titleCode_from_title {
	my $title = shift;

	die "Expected title string as parameter" unless defined($title) and $title;

	my $code = '';

	try {
		require Text::Unidecode;
	}
	catch {
		warn __PACKAGE__ . "::titleCode_from_title() - unable to load 'Text::Unidecode': " . $_;
	};

	try {
		require Text::Undiacritic;
	}
	catch {
		warn __PACKAGE__ . "::titleCode_from_title() - unable to load 'Text::Undiacritic': " . $_;
	};

	my @words = split( /\s+/, $title );
	my @chars = ();
	foreach my $word (@words) {
		my $firstChar = uc( substr( $word, 0, 1 ) );
		if ( $firstChar =~ $Biblio::SICI::Util::TITLE_CODE ) {
			push @chars, $firstChar;
		}
		else {
			try {
				$word = Text::Unidecode::unidecode($word);
			};
			try {
				$word = Text::Undiacritic::undiacritic($word);
			};
			$firstChar = uc( substr( $word, 0, 1 ) );
			if ( $firstChar =~ $Biblio::SICI::Util::TITLE_CODE ) {
				push @chars, $firstChar;
			}
		}
	}

	if ( @chars >= 1 ) {
		$code = join( "", splice( @chars, 0, 6 ) );
		return $code;
	}

	return;
} ## end sub titleCode_from_title

=item STRING C<calculate_check_char>( STRING )

Calculates and returns the check character.
Does B<not> check, if the string param is a valid SICI! 

A check character already present in the passed-in SICI will
simply be ignored.

=cut

sub calculate_check_char {
	my $str = shift;

	return unless defined($str) and $str;

	state $charValues = {
		0 => 0,  1 => 1,  2 => 2,  3 => 3,  4   => 4,  5 => 5,  6 => 6,  7 => 7,
		8 => 8,  9 => 9,  A => 10, B => 11, C   => 12, D => 13, E => 14, F => 15,
		G => 16, H => 17, I => 18, J => 19, K   => 20, L => 21, M => 22, N => 23,
		O => 24, P => 25, Q => 26, R => 27, S   => 28, T => 29, U => 30, V => 31,
		W => 32, X => 33, Y => 34, Z => 35, '#' => 36
	};

	state $valueToChar = { reverse %{$charValues} };

	$str =~ s/\-[0-9A-Z#]\Z/-/o;    # remove check char if present

	my @chars  = split( //, $str );
	my @mapped = ();
	my $i      = 0;
	foreach my $c (@chars) {
		if ( exists $charValues->{$c} ) {
			$mapped[$i] = $charValues->{$c};
		}
		else {
			$mapped[$i] = 36;
		}
		$i++;
	}

	my $sum = 0;
	for ( my $j = $#mapped; $j >= 0; $j -= 2 ) {
		$sum += $mapped[$j];
	}

	$sum *= 3;

	for ( my $j = $#mapped - 1; $j >= 0; $j -= 2 ) {
		$sum += $mapped[$j];
	}

	my $mod = $sum % 37;

	# if remainder == 0; then 0 is the check char
	return '0' if $mod == 0;

	my $checkVal  = 37 - $mod;
	my $checkChar = $valueToChar->{$checkVal};

	return $checkChar;
} ## end sub calculate_check_char

=back

=cut

1;
